text/0088-fix-memory-limitiations-in-session-replays-access-pattern.md

```diff
@@ -0,0 +1,216 @@
+
+**Drawbacks**
+
+- Performance.
+- Requires Snuba work.
+
+**Questions**
+
+### 5. Migrate to CollapsingMergeTree Table Engine and Pre-Aggregate Data in Key-Value Store
+
+**Proposal**
+
+Instead of aggregating in ClickHouse we should maintain a stateful representation of the Replay in an alternative service (such as a key, value store). Additionally, we will migrate to the "VersionedCollapsingMergeTree" table engine and write our progress incrementally.
+
+The versioned engine works as follows:
+
+- A sign is used to determine whether the row is a state row or cancel row.
+- A row can be canceled by re-writing the row with a negative sign.
+- New state rows can be written by incrementing the version and assigning a positive sign.
+- The version and row state must be stored or otherwise fetched from ClickHouse before each write.
+
+The ingestion process can be described as follows:
+
+- A replay-segment is received.
+- The replay-id is used to look up previously aggregated data in the KV store.
+  - If data was returned collate it with the new data and store.
+  - If data was not returned store the new data.
+- The new row is written to ClickHouse with the version column incremented by 1.
+- The old row is re-written to ClickHouse with the sign value set to -1.
+
+**Drawbacks**
+
+- The aggregation process is not expected to be atomic and we will encounter race-conditions where two segments with the same replay-id attempt to mutate the same key at the same time.
+  - To solve this we will need to partition our Kafka messages by replay-id and process sequentially.
+  - This has scalability limitations but those limitations are likely to be less than existing limitations.
+  - Aggregation states can be updated atomicly with some databases. See "Use Alternative OLAP Database" proposal.
+  - Alternatively we can tolerate losing aggregation states and continue using parallel consumers.
+- Kafka will sometimes produce duplicate messages. If we assume order we can set a requirement on segment_id > previous_segment_id.
+  - Order can not be assumed. Valid aggregation states could be lost.
+  - A delay in scheduling in Relay could cause segment_id 1 to be processed prior to segment_id 0.
+  - This is expected to be uncommon but not impossible.
+  - Alternatively, we could tolerate duplicates and accept that segment_count and other values have some margin of error.
+- To eliminate the need for grouping queries we would need to supply the `FINAL` keyword.
+- Our column types are not ideal for this use case.
+
+**Questions**
+
+### 6. Stateful Streaming using Apache Spark or Similar Service
+
+**Proposal**
+
+Instead of aggregating in ClickHouse we should aggregate our replays in a stateful service such as Apache Spark before writing the final result to ClickHouse.
+
+TODO: Josh to fill in specifics.
+
+**Drawbacks**
+
+- Replays are not available until they have finished.
+- There are no existing Apache Spark installations within the Sentry org.
```

Comment:
the datateam manages several spark jobs! cc @trillville

Reply:
That's good to know!  Definitely makes Spark more appealing.